nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2464 @@
+        Self::Update {
+            active_propolis_id: PropolisUuid::from_untyped_uuid(vmm.id),
+            new_sled_id: vmm.sled_id,
+        }
+    }
+}
+
+/// Parameters to the "real" instance update saga.
+#[derive(Debug, Deserialize, Serialize)]
+struct RealParams {
+    serialized_authn: authn::saga::Serialized,
+
+    authz_instance: authz::Instance,
+
+    update: UpdatesRequired,
+
+    orig_lock: instance::UpdaterLock,
+}
+
+const INSTANCE_LOCK_ID: &str = "saga_instance_lock_id";
+const INSTANCE_LOCK: &str = "updater_lock";
+const NETWORK_CONFIG_UPDATE: &str = "network_config_update";
+
+// instance update saga: actions
+
+declare_saga_actions! {
+    instance_update;
+
+    // Become the instance updater
+    BECOME_UPDATER -> "updater_lock" {
+        + siu_become_updater
+        - siu_unbecome_updater
+    }
+
+    // Update network configuration.
+    UPDATE_NETWORK_CONFIG -> "update_network_config" {
+        + siu_update_network_config
+    }
+
+    // Deallocate virtual provisioning resources reserved by the instance, as it
+    // is no longer running.
+    RELEASE_VIRTUAL_PROVISIONING -> "release_virtual_provisioning" {
+        + siu_release_virtual_provisioning
+    }
+
+    // Unassign the instance's Oximeter producer.
+    UNASSIGN_OXIMETER_PRODUCER -> "unassign_oximeter_producer" {
+        + siu_unassign_oximeter_producer
+    }
+
+    // Write back the new instance record, releasing the instance updater lock,
+    // and re-fetch the VMM and migration states. If they have changed in a way
+    // that requires an additional update saga, attempt to execute an additional
+    // update saga immediately.
+    COMMIT_INSTANCE_UPDATES -> "commit_instance_updates" {
+        + siu_commit_instance_updates
+    }
+
+}
+
+// instance update saga: definition
+struct SagaDoActualInstanceUpdate;
+
+impl NexusSaga for SagaDoActualInstanceUpdate {
+    const NAME: &'static str = "instance-update";
+    type Params = RealParams;
+
+    fn register_actions(registry: &mut ActionRegistry) {
+        instance_update_register_actions(registry);
+    }
+
+    fn make_saga_dag(
+        params: &Self::Params,
+        mut builder: DagBuilder,
+    ) -> Result<steno::Dag, SagaInitError> {
+        // Helper function for constructing a constant node.
+        fn const_node(
+            name: impl AsRef<str>,
+            value: &impl serde::Serialize,
+        ) -> Result<steno::Node, SagaInitError> {
+            let value = serde_json::to_value(value).map_err(|e| {
+                SagaInitError::SerializeError(name.as_ref().to_string(), e)
+            })?;
+            Ok(Node::constant(name, value))
+        }
+
+        // Generate a new ID and attempt to inherit the lock from the start saga.
+        builder.append(Node::action(
+            INSTANCE_LOCK_ID,
+            "GenerateInstanceLockId",
+            ACTION_GENERATE_ID.as_ref(),
+        ));
+        builder.append(become_updater_action());
```

Comment:
This is necessary given the "saga actions must be idempotent" constraint. The inherit "inherit lock" operation is required so that, if the action in the start saga that creates the child saga executes again, a second child saga cannot *also* acquire the lock. If we do end up starting two child sagas, only one of them will win the race to compare-and-swap in a new ID, preventing a duplicate saga from running.

As a side note, a duplicate child saga is probably *not* a correctness issue, since all the forward actions should also be idempotent --- but it seems nicer to avoid having to spend a bunch of time serializing all of *that* saga's nodes into the database too when another saga is actually doing the work...

Reply:
Ack, that makes sense!  I wasn't *opposed* to this lock handoff, just curious what merited it, because it did seem like more work.

Using this handoff seems pretty dang explicit + safe!

Reply:
Yeah...it is more work, and I don't know if it is strictly required for correctness, it's more of an optimization to avoid duplicate update sagas. I should document that.

Incidentally, @gjcolombo pointed out at some point that we could probably get away with having the inherit-lock operation just increment the generation number if the updater ID and generation number are equal to those handed off by the previous saga, as a subsequent lock-inherit operation would fail if the generation number has been advanced. I left it the way it is currently, because I thought it was a little bit easier to debug who's holding the lock if the UUID also changes, and I didn't think that it was too much more expensive to be worth trying to optimize out.

I'm open to being convinced otherwise, though! It would be a bit simpler to only have one UUID...